### PR TITLE
Parameter Inconsistency fix

### PR DESCRIPTION
**Pull request by ethancaballero**
This branch converts all global network Variables to ResourceVariable class https://github.com/tensorflow/tensorflow/issues/6360#issuecomment-271741913 by setting `use_resource=True` for global variable_scope to fix the parameter inconsistency issue mentioned in https://github.com/tensorflow/tensorflow/issues/6360#issuecomment-271486091 .

It also puts `opt = tf.train.AdamOptimizer(1e-4)` inside of a scope with `use_resource=True` to (hopefully) prevent updates based on inconsistent Adam momentum and variance accumulator parameters mentioned in https://github.com/openai/universe-starter-agent/issues/67#issuecomment-287343303 .

These changes still needs to A/B tested to see if there's a noticeable positive performance difference. If so, TensorFlow dependency in Readme should be changed to version 1.0 so that future users don't succumb to parameter inconsistency issues.

### CODE REVIEW

**Comment by ethancaballero** [comment]
One needs to use the most recent nightly binary of TF to expose the use_resource parameter.

**Comment by yaroslavvb** [comment]
Curious to see if this makes a difference. From my reading of https://github.com/tensorflow/tensorflow/issues/6360#issuecomment-271741913, it may remove any single variable containing partially updated value. However, the set of Adam parameters as a whole may still be partially updated, since there's no group-level locking

**Comment by ethancaballero** [comment]
I ran this and it only lets one worker thread run, although the one worker thread it runs is slightly more stable than manually running a single worker thread without the `use_resource = True` changes.

**Comment by yaroslavvb** [comment]
What do you mean "more stable", you mean the cost graph has less variance?

**Comment by ethancaballero** [comment]
Average entropy was .1 higher and avg reward rises more slowly but more steadily.

**Comment by tlbtlbtlb** [comment]
I don't understand why resource locking would make any difference with a single worker.

**Comment by jhumplik** [comment]
I started trying this. There are two things that need to be changed. 1. It is necessary to pass ```ps_ops=["Variable", "VariableV2", "VarHandleOp"]``` to the ```tf.train.replica_device_setter()``` since by default it only places ```Variable```, and ```VariableV2``` ops on the ps jobs, and so this code doesn't share any parameters, see:

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/training/device_setter.py#L201

I don't know why ```VarHandleOp``` stands for ```ResourceVariable``` but it seems to be the case. 2. I haven't dug into it yet but there seems to be a change in TF 1.0 and the code as it is places the Adam ops on the worker. Wrapping the ```apply_gradients``` op in a ```tf.device("/job:ps/task:0/cpu:0") ``` context solves this (at least if there's only one ps server). Also the variable scope here https://github.com/ethancaballero/universe-starter-agent/blob/master/a3c.py#L249 is not needed (all the assign ops are created in the ```apply_gradients``` call). 

These changes solve the problem of only one worker running.

I haven't checked for changes in performance when using ```use_resource=True``` yet however I wonder if there is any reason to believe that the functionality described in https://github.com/tensorflow/tensorflow/issues/6360#issuecomment-271741913 has already been implemented? The TODO here suggest it is still work in progress:

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/resource_variable_ops.py#L395
  
Also I second yaroslavvb's comment that even if it was implemented, one still needs a lock on the whole parameter update.

**Comment by yaroslavvb** [comment]
@jhumplik ah, you have better reading comprehension than me, it's not ResourceVariable, but rather a *sister* class to resource variable that will have those semantics, which is not yet implemented
